internal/usecases/table/table_service.go

```diff
@@ -108,39 +121,161 @@ func (ts *TableService) CreateTable(rq *entities.Table) (*response.CreateTableRe
 	}, nil
 }
 
-func (ts *TableService) UpdateTableByID(rq *entities.Table) (*response.UpdateTableResponse, error){
-	exist, err := ts.repo.FindTableByID(rq.ID)
+func (ts *TableService) UpdateTableByID(rq *requests.UpdateTableRequest) (*response.UpdateTableResponse, error) {
+	exist, err := ts.tableRepo.FindTableByID(rq.ID)
 
-	if err != nil{
+	if err != nil && err != gorm.ErrRecordNotFound {
 		return nil, err
 	}
-
-	if exist == nil{
+	if exist == nil {
 		return nil, errors.New("table not found")
 	}
 
-	table, err := ts.repo.UpdateTableByID(rq)
+	c_id, err := utils.StringToUUID(rq.C_ID)
+	if err != nil {
+		return nil, err
+	}
+
+	if exist.Status == "A" && rq.Status != "A" {
+		orderPayload := &entities.Order{
+			ID:   utils.GenerateUUID(),
+			T_ID: exist.ID,
+			Time: time.Now(),
+		}
 
-	if err != nil{
+		order, err := ts.orderRepo.CreateOrderByID(orderPayload)
+		if err != nil {
+			return nil, err
+		}
+
+		m_id, quantity, price, err := ts.parseOrderLine(rq)
+		if err != nil {
+			ts.rollbackOrder(order.ID)
+			return nil, err
+		}
+
+		olinePayload := &entities.OrderLine{
+			ID:       utils.GenerateUUID(),
+			Time:     order.Time,
+			O_ID:     order.ID,
+			M_ID:     *m_id,
+			Quantity: quantity,
+			Price:    float32(price),
+			Url:      rq.OrderLine.Url,
+		}
+
+		oline, err := ts.olineRepo.CreateOrderLine(olinePayload)
+		if err != nil {
+			ts.rollbackOrder(order.ID)
+			return nil, err
+		}
+
+		return ts.updateTableWithOrderNOrderLine(order, oline, exist, c_id, rq)
+	}
+
+	return ts.updateTable(rq, c_id)
+}
+
+func (ts *TableService) rollbackOrder(id pgtype.UUID) {
+	if err := ts.orderRepo.DeleteOrderByID(id); err != nil {
+		panic("Can't roll back order in booking table")
+	}
+}
+
+func (ts *TableService) rollbackOrderLine(id pgtype.UUID) {
+	if err := ts.olineRepo.DeleteOrderLineByID(id); err != nil {
+		panic("Can't roll back order in booking table")
+	}
+}
+
+func (ts *TableService) parseOrderLine(rq *requests.UpdateTableRequest) (*pgtype.UUID, int, float64, error) {
+	m_id, err := utils.StringToUUID(rq.OrderLine.M_ID)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	quantity, err := strconv.Atoi(rq.OrderLine.Quantity)
+	if err != nil {
+		return nil, 0, 0, err
+	}
+
+	price, err := strconv.ParseFloat(rq.OrderLine.Price, 64)
```

Comment:
Why parse the price as a float64? since The OrderLine struct uses float32? (ref from line 157)

Reply:
Return type of this function Is float 64 by default. I don't think I can return float 32

Reply:
You can cast the value back to float32, but I still don't know about the accuracy of the value since casting can lead to precision loss because we have to compute this value. If you find it okay so am I 555555555555.